src/test/java/org/opensearch/security/dlic/rest/api/TenantInfoActionTest.java

```diff
@@ -72,17 +79,78 @@ public void testTenantInfoAPIUpdate() throws Exception {
+                "[{\"op\": \"add\",\"path\": \"/config/dynamic/kibana/opendistro_role\","
+                        + "\"value\": \"opendistro_security_internal\"}]",
+                new Header[0]);
         Assert.assertEquals(HttpStatus.SC_OK, response.getStatusCode());
 
-        response = rh.executePutRequest(BASE_ENDPOINT + "/api/rolesmapping/opendistro_security_internal", payload, new Header[0]);
+        response = rh.executePutRequest(BASE_ENDPOINT + "/api/rolesmapping/opendistro_security_internal", payload,
+                new Header[0]);
         Assert.assertEquals(HttpStatus.SC_OK, response.getStatusCode());
 
         rh.sendAdminCertificate = false;
         response = rh.executeGetRequest(ENDPOINT);
         Assert.assertEquals(HttpStatus.SC_OK, response.getStatusCode());
     }
+
+    @Test
+    public void testParallelPutRequests() throws Exception {
+
+        setup();
+
+        rh.keystore = "restapi/kirk-keystore.jks";
+        rh.sendAdminCertificate = true;
+
+        String[] tenantNames = { "tenant_1", "tenant_2" };
+        String[] descriptions = { "create tenant 1", "create tenant 2" };
+        HttpResponse[] responses = executeMultipleAsyncPutRequest(tenantNames, descriptions);
+        int numCreatedResponses = 0;
+        int numConflictResponses = 0;
+        Integer requestIndexCreated = null;
+        for (int i = 0; i < responses.length; i++) {
+            HttpResponse response = responses[i];
+            if (response.getStatusCode() == HttpStatus.SC_CREATED) {
+                numCreatedResponses++;
+                requestIndexCreated = i;
+            } else if (response.getStatusCode() == HttpStatus.SC_CONFLICT) {
+                numConflictResponses++;
+            }
+        }
+        Assert.assertEquals(1, numCreatedResponses); // check that no more than 1 created status is read
+        Assert.assertEquals(1, numConflictResponses); // check that no more than 1 conflict status is read
+        Assert.assertNotNull(requestIndexCreated); // check that a created status is read
+
+        HttpResponse response = rh
+                .executeGetRequest(BASE_ENDPOINT + "/api/tenants/" + tenantNames[requestIndexCreated]);
+        Assert.assertEquals(descriptions[requestIndexCreated],
+                response.findValueInJson(tenantNames[requestIndexCreated] + ".description"));
+    }
+
+    private HttpResponse[] executeMultipleAsyncPutRequest(final String[] names,
+            final String[] descriptions) throws Exception {
+        final int numOfRequests = Math.min(names.length, descriptions.length);
+        final String TENANTS_ENDPOINT = BASE_ENDPOINT + "/api/tenants/";
+        final ExecutorService executorService = Executors.newFixedThreadPool(numOfRequests);
+        try {
+            List<Future<HttpResponse>> futures = new ArrayList<>(numOfRequests);
+            for (int i_ = 0; i_ < numOfRequests; i_++) {
+                final int i = i_;
```

Comment:
why need a final declaration here?

Reply:
I previously got a weird java warning asking me to use that, but it looks like it's not required with the current version of this function, resolved!